smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/util/StringStore.java

```diff
@@ -0,0 +1,131 @@
+ * form of compression on long protocol serde files.
+ */
+public class StringStore {
+    private final ConcurrentHashMap<String, String> literalToVariable = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, String> variableToLiteral = new ConcurrentHashMap<>();
+    private final Set<String> writelog = new HashSet<>();
+
+    /**
+     * @param literal - a literal string value.
+     * @return the variable name assigned for that string, which may have been encountered before.
+     */
+    public String var(String literal) {
+        if (literal == null) {
+            throw new RuntimeException("Literal must not be null.");
+        }
+        if (literalToVariable.containsKey(literal)) {
+            return literalToVariable.get(literal);
+        }
+        literalToVariable.put(literal, this.assignKey(literal));
+        return literalToVariable.get(literal);
+    }
+
+    /**
+     * Outputs the generated code for any constants that have been
+     * allocated but not yet retrieved.
+     */
+    public String getIncremental() {
+        TreeMap<String, String> map = new TreeMap<>(variableToLiteral);
+
+        StringBuilder sourceCode = new StringBuilder();
+        Set<String> incrementalKeys = new HashSet<>();
+
+        for (Map.Entry<String, String> entry : map.entrySet()) {
+            String v = entry.getKey();
+            String l = entry.getValue();
+            if (writelog.contains(v)) {
+                // sourceCode.append(String.format("// const %s = \"%s\";%n", v, l));
+            } else {
+                incrementalKeys.add(v);
+                sourceCode.append(String.format("const %s = \"%s\";%n", v, l));
+            }
+        }
+
+        writelog.addAll(incrementalKeys);
+
+        return sourceCode.toString();
+    }
+
+    /**
+     * Assigns a new variable or returns the existing variable for a given string literal.
+     */
+    private String assignKey(String literal) {
+        if (literalToVariable.containsKey(literal)) {
+            return literalToVariable.get(literal);
+        }
+        String variable = allocateVariable(literal);
+        variableToLiteral.put(variable, literal);
+        literalToVariable.put(literal, variable);
+        return variable;
+    }
+
+    /**
+     * Assigns a unique variable using the letters from the literal.
+     * Prefers the uppercase or word-starting letters.
+     */
+    private String allocateVariable(String literal) {
+        String[] sections = literal.split("[-_\\s]");
+        StringBuilder v = new StringBuilder("_");
+        Queue<Character> deconfliction = new LinkedList<>();
+        if (sections.length > 1) {
+            Arrays.stream(sections)
+                .map(s -> s.charAt(0))
+                .filter(this::isAllowedChar)
+                .forEach(v::append);
+        } else {
+            for (int i = 0; i < literal.length(); ++i) {
```

Comment:
is this not skipping the first character because you used `++i` rather than `i++`?

Reply:
the 3rd part of the for is not evaluated until the iteration completes, so it shouldn't make a difference. 

this is an old habit from hearing that pre-increment is (imperceptibly) faster than post-increment 😅

Reply:
if it doesn't matter either way, I'd recommend ++i to not trip up people in the future reading this code. I'd assume it's important to start at 1 here, but it doesn't seem to be the case.